### DISCUSSION TITLE
Sound speed and pitch in as3

### DISCUSSION THREAD

**Discussion by raresn** [airsdk/Adobe-Runtime-Support#2565] [Guides & Tutorials]
Hey guys! We have a new challenge ahead :)
We need to slow down/speed up a sound without changing it's pitch.
Does someone know how to do that? Or has any libraries/anes that might help? 
We've been searching like crazy and nothing....

Thank you!

**Comment by VCone** [comment]
You could try: <br>
http://iq12.com/old_blog/2009/08/25/real-time-pitch-shifting/

PS: Better to ask coding question on StackOverflow under the AS3 tag. Then showcase your result on this Github.

**Reply by raresn** [reply]
Good tip with StackOverflow - thanks. 
And thank you for the project link. After running it it doesn't really do what i expected it to do - speed up the sound without changing the pitch.

**Reply by VCone** [reply]
My bad. After re-reading my own [answer on S.O site](https://stackoverflow.com/a/36158808/2057709) I realise misread "whilst adjusting pitch" as saying "whilst keeping pitch". Don't ask how I missed it when it's there in **bold** letters for extra noticing. 

PS: Is the code usable as a starting point? If yes maybe read up on the logic of keeping pitch then ask something on StackOverflow for further advice if stuck. (note: You need to have some code attempt before they can help to fix anything. Otherwise they'll downvote you instead).

**Reply by ajwfrost** [reply]
Presumably you'd need to combine the pitch-shifting with re-sampling? Or to put it another way: if you want to speed up a sound then you need to either play it back at a higher rate, or resample it to have fewer samples so that it plays more quickly. But doing that would affect the pitch; so then you can use that algorithm to shift the pitch back down again.

Don't know if there's any library to do both at once. I would suggest that a C (or similar) library could be used for this and then turned into a SWC library with the help of crossbridge - that's likely to be more efficient than trying to cross-compile/transcribe it manually.

**Reply by raresn** [reply]
@VCone  - no worries. Yes i wrote the part where i can speed the sound up, but the pitch is super bad.
@ajwfrost - the last part with a C library is the conclusion i ended up with after talking with chatGPT till 3 am last night :))

**Comment by teotigraphix** [comment]
Rubber Band (Pitch/Time stretch algorithms) could easily be made into an ANE for Android/Windows etc.

https://breakfastquay.com/rubberband/why.html

**Reply by raresn** [reply]
This actually sounds very cool! If it's just C++ maybe i can turn it into a SWC with Crossbridge like ajwfrost suggested?

**Reply by teotigraphix** [reply]
Actually, I don't have that answer. I am knee deep in my own audio stuff atm but I was thinking about that down the road in AS3. :)

With an ANE you just need some AS3 API to access the very simple one class C++ API Rubber Band uses. 

Caustic, an Android app uses this library for time/ptich stretching since around 2015 on Mac/Win/iOS/Android and it works well. Real time wasn't tried back then but it could probably work on most newer devices now.

**Comment by 2jfw** [comment]
Not sure if this meets all your requirements but have a look at:
https://github.com/Adolio/Syrinx-Sound-Manager

I think its the most complete and up-to-date as3 project regarding sounds and effects etc..

**Reply by raresn** [reply]
Just might be it! I will try it and let you know! Thanks!

**Reply by raresn** [reply]
@2jfw 
Here is what i found:  I am trying to change the pitch independent from the speed. 
Meaning, i am trying to speed up a sound but adjust the pitch so it doesn't sound like a cartoon :)
The example there, increases the speed, therefore increasing the pitch.
